### ISSUE TITLE
Error with 4.40.2013_08_greengenes_97_otus.with_euks.spkg

### ISSUE THREAD

**Issue by michoug** [wwood/singlem#26]
Hi,
When I tried to run the following command

`singlem pipe --sequences *R1.fastq.gz --otu_table test16S.csv --singlem_packages ~/singlem_extra_packages/release1/4.40.2013_08_greengenes_97_otus.with_euks.spkg --threads 5 --debug`

I got the following error for one sample

```
08/07/2018 12:54:09 PM INFO: Working on S_16_H0527_N7_F1_H11_L007_R1_hits
08/07/2018 12:54:09 PM DEBUG: Running nucleotide pipeline
08/07/2018 12:54:09 PM DEBUG: Using 1 HMMs to search
08/07/2018 12:54:09 PM DEBUG: Detected file format FORMAT_FASTA
08/07/2018 12:54:09 PM DEBUG: raw read unpacking command chunk: cat /dev/shm/tmp1S4q_m/graftm_nucleotide_search/S_16_H0527_N7_F1_H11_L007_R1_hits.fa
08/07/2018 12:54:09 PM DEBUG: Running command: cat /dev/shm/tmp1S4q_m/graftm_nucleotide_search/S_16_H0527_N7_F1_H11_L007_R1_hits.fa | nhmmer --incE 1e-05 -E 1e-05 --cpu 1 -o /dev/null --noali --tblout /dev/shm/tmp1S4q_m/graftm_separates/4.40.2013_08_greengenes_97_otus.with_euks/S_16_H0527_N7_F1_H11_L007_R1_hits/S_16_H0527_N7_F1_H11_L007_R1_hits.hmmout.txt /home/michoug/singlem_extra_packages/release1/4.40.2013_08_greengenes_97_otus.with_euks.spkg/4.40.2013_08_greengenes_97_otus.with_euks/graftm6wK98G_search.hmm -
08/07/2018 12:54:09 PM DEBUG: Running extern cmd: cat /dev/shm/tmp1S4q_m/graftm_nucleotide_search/S_16_H0527_N7_F1_H11_L007_R1_hits.fa | nhmmer --incE 1e-05 -E 1e-05 --cpu 1 -o /dev/null --noali --tblout /dev/shm/tmp1S4q_m/graftm_separates/4.40.2013_08_greengenes_97_otus.with_euks/S_16_H0527_N7_F1_H11_L007_R1_hits/S_16_H0527_N7_F1_H11_L007_R1_hits.hmmout.txt /home/michoug/singlem_extra_packages/release1/4.40.2013_08_greengenes_97_otus.with_euks.spkg/4.40.2013_08_greengenes_97_otus.with_euks/graftm6wK98G_search.hmm -
08/07/2018 12:55:30 PM ERROR: Multiple sequences found with the same ID. The input sequences are either ill formated or are interleaved. If you provided GraftM with an interleaved sequence file, please split them into forward and reverse reads, and provide to the the appropriate flags (--forward, --reverse). Otherwise, it appears that you have provided sequences with redundant IDs. GraftM doesn't know how to deal with these, so please remove/rename sequences with duplicate keys.
Traceback (most recent call last):
  File "/home/michoug/miniconda2/bin/graftM", line 409, in <module>
    Run(args).main()
  File "/home/michoug/miniconda2/lib/python2.7/site-packages/graftm/run.py", line 588, in main
    self.graft()
  File "/home/michoug/miniconda2/lib/python2.7/site-packages/graftm/run.py", line 398, in graft
    self.args.evalue
  File "/home/michoug/miniconda2/lib/python2.7/site-packages/graftm/timeit.py", line 10, in timed
    result = method(*args, **kw)
  File "/home/michoug/miniconda2/lib/python2.7/site-packages/graftm/sequence_searcher.py", line 1021, in nt_db_search
    hit_reads_fasta)
  File "/home/michoug/miniconda2/lib/python2.7/site-packages/graftm/sequence_searcher.py", line 1095, in search_and_extract_nucleotides_matching_nucleotide_database
    hits
  File "/home/michoug/miniconda2/lib/python2.7/site-packages/graftm/sequence_searcher.py", line 506, in _extract_from_raw_reads
    complement_info = self._extract_multiple_hits(hits, tmp.name, output_path)  # split them into multiple reads
  File "/home/michoug/miniconda2/lib/python2.7/site-packages/graftm/sequence_searcher.py", line 432, in _extract_multiple_hits
    raise InterleavedFileError()
graftm.sequence_searcher.InterleavedFileError
STDOUT was:
                             GraftM 0.11.1

                                GRAFT

                       Joel Boyd, Ben Woodcroft

                                                         __/__
                                                  ______|
          _- - _                         ________|      |_____/
           - -            -             |        |____/_
           - _     >>>>  -   >>>>   ____|
          - _-  -         -             |      ______
             - _                        |_____|
           -                                  |______

```
Any ideas ? the command with the default set worked flawlessly

**Comment by wwood** [comment]
Hmm, strange. Would you mind rerunning it with --working_directory, and then sending me folder compressed somehow please? e.g. via email or dropbox etc?

Thanks, ben

**Comment by michoug** [comment]
[Test.zip](https://github.com/wwood/singlem/files/2269187/Test.zip)
[logTest.txt](https://github.com/wwood/singlem/files/2269188/logTest.txt)

Here it is

**Comment by wwood** [comment]
Thanks. Just to let you know though I'm currently on holidays so it will take a little time to get to it.

**Comment by wwood** [comment]
Hi,
I'm not sure why, but it appears you do actually have multiple reads with the same ID in them e.g.
```
>K00235:30:HG3N2BBXX:7:2109:26304:18826 1:N:0:CATACGGA
GTGAATACGTTCCCGGGCCTTGTACACACCGCCCGTCACGCCATGGAGTCGAGTTGCAGACTCCAATCCGAACTGGGGCCGGTTTTTATGGATTGGCTTCCCCTCGCGGGTTCGCGACCCTTTGTACCGGCCATTGTAACACGTGTGTAGC
>K00235:30:HG3N2BBXX:7:2109:26304:18826 1:N:0:CATACGGA
GCTACACACGTGTTACAATGGCCGGTACAAAGGGTCGCGAACCCGCGAGGGGAAGCCAATCCATAAAAACCGGCCCCAGTTCGGATTGGAGTCTGCAACTCGACTCCATGGCGTGACGGGCGGTGTGTACAAGGCCCGGGAACGTATTCAC
```
I'm not sure whether that was intentional on your part, but the fix would just be to rename the sequences e.g. using this script https://github.com/wwood/bbbin/blob/master/sequenceNamer.pl
```
zcat S_16_H0527_N7_F1_H11_L007_R1.fastq.gz |fq2fa.sh |grep -v '>' |sequenceNamer.pl |pigz >S_16_H0527_N7_F1_H11_L007_R1.renamed.fna.gz
```
where
```
alias fq2fa='awk '\''{print ">" substr($0,2);getline;print;getline;getline}'\'''
```
Quality is ignored by singlem and graftm.

**Comment by michoug** [comment]
Hum, strange, I'll try this, thanks!
However, don't understand why it fails for this greegenes package and succeeds for the default proteins?

**Comment by wwood** [comment]
Probably because most reads have unique names, I suppose

**Comment by wwood** [comment]
Let me know if there's any thing more to do here. Thanks.

**Comment by wwood** [comment]
That is quite odd. What happens when you do this on the working directory _hits file
```
grep '>' renamed_hits.fa |sort |uniq -c |sort -rn |head
```
Are there any dups?

**Comment by michoug** [comment]
No they aren't any

**Comment by wwood** [comment]
Hmm, can you send the working directory and log again please?

**Comment by michoug** [comment]
[Test.zip](https://github.com/wwood/singlem/files/2289984/Test.zip)
[Test.txt](https://github.com/wwood/singlem/files/2289985/Test.txt)

Here they are

**Comment by wwood** [comment]
Hmm..
```
$ grep '>' Test/graftm_nucleotide_search/S_16_H0527_N7_F1_H11_L007_R1.renamed_hits.fa |sort |uniq -c |sort -rn |head
      2 >seq2452201
      2 >seq2451145
      2 >seq18975201
      1 >seq9991085
      1 >seq9982357
      1 >seq9980476
      1 >seq9980356
      1 >seq9973186
      1 >seq9971690
      1 >seq9969550
```

**Comment by wwood** [comment]
So they aren't duped in the original fna?

**Comment by michoug** [comment]
No, I confirmed, I ran with just 33 sequences including the ones that are indicated as duplicate in the renamed_hits.fa and ran into the same issue.
Here are the initial files and log

[SubsetLog.txt](https://github.com/wwood/singlem/files/2290357/SubsetLog.txt)
[Subset.zip](https://github.com/wwood/singlem/files/2290358/Subset.zip)
[subset_fasta.txt](https://github.com/wwood/singlem/files/2290362/subset_fasta.txt)

**Comment by wwood** [comment]
It appears this was caused by the rare situation where the a read matches in both directions. I've fixed this in the above commit in the dev branch, which is fine for read input (it chooses the direction that has the highest bitscore for HMM alignment). However, there's a corner case for longer sequences like contigs where they legitimately have 2 hits, one in each direction. That will require more time than I have right now, follow #27 if you are interested.

Thank you for your clear and patient reporting of this bug - really helped me nail it down.
